allennlp/models/encoder_decoders/nlvr_semantic_parser.py

```diff
@@ -0,0 +1,365 @@
+                 action_embedding_dim: int,
+                 attention_function: SimilarityFunction) -> None:
+        # Setting num_entity_types to 0 since that is needed only for linking, and we
+        # will not use linking in NLVR (yet).
+        super(NlvrDecoderStep, self).__init__(encoder_output_dim=encoder_output_dim,
+                                              action_embedding_dim=action_embedding_dim,
+                                              attention_function=attention_function,
+                                              num_entity_types=0)
+
+    @classmethod
+    def _compute_new_states(cls,  # type: ignore
+                            state: NlvrDecoderState,
+                            log_probs: torch.Tensor,
+                            hidden_state: torch.Tensor,
+                            memory_cell: torch.Tensor,
+                            action_embeddings: torch.Tensor,
+                            attended_question: torch.Tensor,
+                            considered_actions: List[List[int]],
+                            allowed_actions: List[Set[int]],
+                            max_actions: int = None) -> List[NlvrDecoderState]:
+        """
+        We override the ``_compute_new_states`` method here to compute new states based on coverage
+        loss.  For each action in considered actions, if it is in the agenda, we update the checklist
+        and recalculate the score. We'll have to iterate through considered actions in either case
+        because we have to update action histories anyway.
+        """
+        # TODO(pradeep): We do not have a notion of ``allowed_actions`` for NLVR for now, but this
+        # may be applicable in the future.
+        # Note that we're not sorting these probs. We'll score checklists and sort states based on
+        # checklist scores later.
+        probs = torch.exp(log_probs)
+        # batch_index -> [(group_index, action_index, action, checklist, score)]
+        next_states_info: Dict[int, List[Tuple[int, int, int, Variable, Variable]]] = defaultdict(list)
+        for group_index, (batch_index, instance_action_probs) in enumerate(zip(state.batch_indices, probs)):
+            instance_agenda = state.agenda[group_index]
+            instance_agenda_mask = state.agenda_mask[group_index]
+            instance_checklist = state.checklist[group_index]
+            # action_prob is a Variable.
+            for action_index, action_prob in enumerate(instance_action_probs):
+                # This is the actual index of the action from the original list of actions.
+                action = considered_actions[group_index][action_index]
+                if action == -1:
+                    # Ignoring padding.
+                    continue
+                # If action is not in instance_agenda, mask_variable, and checklist_addition will be
+                # all 0s.
+                checklist_mask = instance_agenda.eq(action)
```

Comment:
What are the shapes of these things?  I'm having a hard time convincing myself that this logic is correct.

Reply:
They're all as long as the padded agenda. Added a comment.